### DISCUSSION TITLE
printing the log directly will obviously cause a JVM deadlock

### DISCUSSION THREAD

**Discussion by hustchaya** [async-profiler/async-profiler#821] [Q&A] [answered]
In the recordSample function, printing the log directly will obviously cause a JVM deadlock. Would you like to inquire about this detailed principle and solution?

**Comment by apangin** [comment, marked as answer]
I'm not sure I understand the question. Are you talking about an existing async-profiler issue or you are trying to develop new functionality?

Since logging relies on `*printf` functions, it's unsafe to use it inside a signal handler.
See [signal-safety(7)](https://man7.org/linux/man-pages/man7/signal-safety.7.html) for the list of functions safe to call inside a signal handler.

**Reply by hustchaya** [reply]
I'm trying to make some modifications that suit my product. Thank you very much for your reply. My doubts have been resolved.
